modules/apps/site/site-cms-site-initializer/src/main/resources/META-INF/resources/js/main/components/forms/PermissionsTable.tsx

```diff
@@ -0,0 +1,387 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2025 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+import {Body, Cell, Head, Row, Table, Text} from '@clayui/core';
+import {ClayCheckbox, ClaySelect} from '@clayui/form';
+import React, {useMemo, useState} from 'react';
+
+interface PermissionItem {
+	actionIds: string[];
+	roleName: string;
+}
+
+/**
+ * Returns a new permissions object with the particular actionId at
+ * a certain index toggled. Called when clicking on a checkbox in the
+ * permissions table.
+ *
+ * @param permissions
+ * @param actionId
+ * @param index
+ * @returns PermissionItem[]
+ */
+const toggleActionIdAtIndex = (
+	permissions: PermissionItem[],
+	actionId: string,
+	index: number
+) => {
+	const newPermissions = permissions.slice();
+
+	newPermissions[index].actionIds = permissions[index].actionIds.includes(
+		actionId
+	)
+		? permissions[index].actionIds.filter((item) => item !== actionId)
+		: [...permissions[index].actionIds, actionId];
+
+	return newPermissions;
+};
+
+/**
+ * Returns a new permissions object with all the actionIds enabled
+ * for the listed roles, while the rest of the roles have the actionIds
+ * disabled. Called when selecting from the 'Viewable By' dropdown.
+ *
+ * @param permissions
+ * @param roleNames
+ * @param actionIdsToEnable
+ * @returns PermissionItem[]
+ */
+const enableActionIdsForRoles = (
+	permissions: PermissionItem[],
+	roleNamesToEnable: string[],
+	actionIdsToEnable: string[]
+) => {
```

Comment:
It's good practice to add return types to these functions so the return value is typed as well. Mainly in this function's case since it's just returning a new untyped object.

```
const enableActionIdsForRoles = (
	permissions: PermissionItem[],
	roleNamesToEnable: string[],
	actionIdsToEnable: string[]
) : PermissionItem[] => {
```

Reply:
Ooo, good practice, will include the typed return value. Thanks for the snippet too!